src/phase0/state_transition/epoch_processing.rs

```diff
@@ -254,8 +254,25 @@ pub fn process_effective_balance_updates<
         MAX_VALIDATORS_PER_COMMITTEE,
         PENDING_ATTESTATIONS_BOUND,
     >,
-    _context: &Context,
+    context: &Context,
 ) -> Result<(), Error> {
+    // Update effective balances with hysteresis
+    for i in 0..state.validators.len() {
+        let validator = &mut state.validators[i];
+        let balance = &state.balances[i];
+        let hysteresis_increment =
+            context.effective_balance_increment / context.hysteresis_quotient;
+        let downward_threshold = hysteresis_increment * context.hysteresis_downward_multiplier;
+        let upward_threshold = hysteresis_increment * context.hysteresis_upward_multiplier;
+        if balance + downward_threshold < validator.effective_balance
+            || validator.effective_balance + upward_threshold < *balance
```

Comment:
General rust question here within the `if` statement where `balance` is a `&`:
1. I didn't have to dereference `balance` in `balance + downward_threshold < validator.effective_balance`
2. However, I did have to dereference `balance` in `validator.effective_balance + upward_threshold < *balance`

Is this because comparison operators will compare expressions of only the same types? 

i.e., the statement in (1) somehow converts the LHS `balance + downward_threshold` to a normal `u64` and compares that to the RHS, also a `u64`. And in (2), if balance is not dereferenced, then the LHS is a `u64` and RHS is `&u64`, which can't be compared.

Reply:
you just need `PartialOrd` implemented to give meaning to operators like `<`

but unless you have something like `PartialOrd<&T> for T` then `rustc` won't know how to handle `a < b` where `a: &T` and `b: T`.

`.` has some special magic for auto-dereferencing and taken together these explain what you see.